[Core][Metrics] Add Seatunnel Metrics module

lvlv-feifei opened the pull request:
<!--

Thank you for contributing to SeaTunnel! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

## Contribution Checklist

  - Make sure that the pull request corresponds to a [GITHUB issue](https://github.com/apache/incubator-seatunnel/issues).

  - Name the pull request in the form "[Feature] [component] Title of the pull request", where *Feature* can be replaced by `Hotfix`, `Bug`, etc.

  - Minor fixes should be named following this pattern: `[hotfix] [docs] Fix typo in README.md doc`.

-->
https://github.com/apache/incubator-seatunnel/issues/2735
https://github.com/apache/incubator-seatunnel/issues/2592
## Purpose of this pull request
Complete the code of seatuunel-metrics according to the framework of the issue, and add module seatuunel-metrics, including seatuunel-core, seatuunel-flink, seatuunel-spark three modules.
seatuunel-core is responsible for the basic definition of the seatunnel indicator system, including the definition of indicators and external interfaces
seatuunel-flink is responsible for the connection between seatunel and flink metrics, and accepts flink metrics
seatuunel-spark is responsible for the connection between seatuunel and spark metrics, and accepts spark metrics

<!-- Describe the purpose of this pull request. For example: This pull request adds checkstyle plugin.-->

## Check list

* [x] Code changed are covered with tests, or it does not need tests for reason:
* [x] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/incubator-seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [x] If necessary, please update the documentation to describe the new feature. https://github.com/apache/incubator-seatunnel/tree/dev/docs

lvlv-feifei committed:
seatunnel-metrics init commit

lvlv-feifei committed:
Merge branch 'apache:dev' into dev

lvlv-feifei committed:
seatunnel-metrics config add

lvlv-feifei committed:
Merge branch 'apache:dev' into dev

Hisoka-X commented:
Please solved ci problems. Thanks

Hisoka-X commented on seatunnel-apis/seatunnel-api-flink/src/main/java/org/apache/seatunnel/flink/FlinkEnvironment.java:
```diff
@@ -322,4 +326,61 @@ private void setCheckpoint() {
+        if(!config.hasPath(ConfigKeyName.Metric_Class)){
+            return StreamExecutionEnvironment.getExecutionEnvironment();
+        }
+        //构建flink-metrics参数
```
Use english only

Hisoka-X commented on seatunnel-apis/seatunnel-api-flink/src/main/java/org/apache/seatunnel/flink/FlinkEnvironment.java:
```diff
@@ -322,4 +326,61 @@ private void setCheckpoint() {
         }
     }
 
+    public StreamExecutionEnvironment creatMetricStreamEEnvironment() {
```
```suggestion
    public StreamExecutionEnvironment creatStreamEnvironment() {
```
Not only have metrics config will run this method.

Hisoka-X commented on seatunnel-apis/seatunnel-api-flink/src/main/java/org/apache/seatunnel/flink/util/ConfigKeyName.java:
```diff
@@ -45,4 +45,9 @@ private ConfigKeyName() {
     public static final String MIN_STATE_RETENTION_TIME = "execution.query.state.min-retention";
     public static final String STATE_BACKEND = "execution.state.backend";
     public static final String PLANNER = "execution.planner";
+    public static final String Metric_Interval = "execution.metrics.interval";
```
Use unified name in spark/flink. There config name is for `SeaTunnel`, not only for `Flink` or `Spark`.

Hisoka-X commented on seatunnel-apis/seatunnel-api-flink/src/main/java/org/apache/seatunnel/flink/FlinkEnvironment.java:
```diff
@@ -322,4 +326,61 @@ private void setCheckpoint() {
+                        .stringType()
+                        .noDefaultValue();
+
+        Configuration seatunnel_reporter = new Configuration().set(REPORTERS_LIST, "seatunnel_reporter").set(REPORTER_CLASS, "org.apache.seatunnel.metrics.flink.SeatunnelMetricReporter");
```
use `lower camel case` for field and parameter. Change `seatunnel_reporter` to `seatunnelReporter`

Hisoka-X commented on seatunnel-apis/seatunnel-api-flink/src/main/java/org/apache/seatunnel/flink/FlinkEnvironment.java:
```diff
@@ -322,4 +326,61 @@ private void setCheckpoint() {
+            return StreamExecutionEnvironment.getExecutionEnvironment();
+        }
+        //构建flink-metrics参数
+        ConfigOption<String> REPORTERS_LIST =
```
Maybe you can create an new method called `initMetricConfig(Configuration config)` to cover this logic.

Hisoka-X commented on seatunnel-apis/seatunnel-api-spark/src/main/java/org/apache/seatunnel/spark/SparkEnvironment.java:
```diff
@@ -178,6 +179,22 @@ public static <T extends Object> T sinkProcess(SparkEnvironment environment, Bas
     }
+
+    private SparkSession.Builder creatMetricBuilder(SparkSession.Builder builder){
+        if(config.hasPath("spark.metrics.class")){
```
Same problem, you should use  unified name

Hisoka-X commented on seatunnel-examples/seatunnel-flink-examples/src/main/resources/examples/fake_to_console.conf:
```diff
@@ -23,6 +23,11 @@ env {
   execution.parallelism = 1
   #execution.checkpoint.interval = 10000
   #execution.checkpoint.data-uri = "hdfs://localhost:9000/checkpoint"
+  execution.metrics.class = org.apache.seatunnel.metrics.core.reporter.PrometheusPushGatewayReporter
```
Not add this in default config, because other developer environment don't have Prometheus.

Hisoka-X commented on seatunnel-examples/seatunnel-spark-examples/src/main/resources/examples/spark.batch.conf:
```diff
@@ -27,6 +27,10 @@ env {
   spark.executor.cores = 1
   spark.executor.memory = "1g"
   spark.master = local
+  spark.metrics.class = org.apache.seatunnel.metrics.core.reporter.PrometheusPushGatewayReporter
```
Same problems. You can use comment to show this config.

Hisoka-X commented on seatunnel-metrics/seatunnel-metrics-core/pom.xml:
```diff
@@ -0,0 +1,26 @@
+
+    <dependencies>
+        <dependency>
+            <groupId>io.prometheus</groupId>
```
Core can't contains prometheus, your should create new module call `seatunnel-metrics-prometheus`, then implement your metrics push to prometheus logic in there. The core should only have interface for metrics and common code.

Hisoka-X commented on seatunnel-metrics/seatunnel-metrics-core/src/main/java/org/apache/seatunnel/metrics/core/reporter/PrometheusPushGatewayReporter.java:
```diff
@@ -0,0 +1,238 @@
+/**
+ * A reporter which outputs measurements to PrometheusPushGateway
+ */
+public class PrometheusPushGatewayReporter implements MetricReporter {
```
Move to `seatunnel-metrics-prometheus`

Hisoka-X commented on seatunnel-metrics/seatunnel-metrics-core/src/main/java/org/apache/seatunnel/metrics/core/MetricInfo.java:
```diff
@@ -0,0 +1,52 @@
+        builder.append("helpString: ")
+                .append(this.helpString)
+                .append(lineSeparator);
+        for(int i=0;i<this.dimensionKeys.size();i++){
```
Code need be formated.

Hisoka-X commented on seatunnel-metrics/seatunnel-metrics-flink/src/main/java/org/apache/seatunnel/metrics/flink/SeatunnelMetricReporter.java:
```diff
@@ -0,0 +1,104 @@
+        jobName = config.getString("jobName","flinkJob");
+        //config.
+        //String string = metricConfig.getString("name", "de");
+        //log.info("StreamMetricReporter init:{}", string);
```
Remove useless code.

Hisoka-X commented on seatunnel-metrics/seatunnel-metrics-flink/src/main/java/org/apache/seatunnel/metrics/flink/SeatunnelMetricReporter.java:
```diff
@@ -0,0 +1,104 @@
+ * exports Flink metrics to Seatunnel
+ */
+public class SeatunnelMetricReporter extends AbstractSeatunnelReporter implements Scheduled {
+    private static final Logger log = LoggerFactory.getLogger(SeatunnelMetricReporter.class);
```
```suggestion
    private static final Logger LOGGER = LoggerFactory.getLogger(SeatunnelMetricReporter.class);
```

lvlv-feifei committed:
codeStyle update

lvlv-feifei committed:
codeStyle update again

lvlv-feifei committed:
codeStyle seatunnel-spark update

TaoZex committed:
[Imporve][Connector-V2]Parameter verification for connector V2 kafka sink (#2866)

* parameter verification

* update

* update

liugddx committed:
[Improve][DOC] Perfect the connector v2 doc (#2800)

* [Improve][DOC] Perfect the connector v2 doc

* Update seatunnel-connectors-v2/README.zh.md

Co-authored-by: Hisoka <fanjiaeminem@qq.com>

* [Improve][DOC] A little tinkering

* [Improve][DOC] A little tinkering

* [Doc][connector] add Console sink doc

close #2794

* [Doc][connector] add Console sink doc

close #2794

* fix some problem

* fix some problem

* fine tuning

Co-authored-by: Hisoka <fanjiaeminem@qq.com>

EricJoy2048 committed:
add seatunnel-examples from gitignore (#2892)

zhaomin1423 committed:
[Improve][connector-jdbc] Calculate splits only once in JdbcSourceSplitEnumerator (#2900)

hailin0 committed:
[Bug][Connector-V2] Fix wechat sink data serialization (#2856)

Hisoka-X committed:
[Improve][Connector-V2] Improve orc write strategy to support all data types (#2860)

* [Improve][Connector-V2] Improve orc write strategy to support all data types

Co-authored-by: tyrantlucifer <tyrantlucifer@gmail.com>

hailin0 committed:
[Bug][seatunnel-translation-base] Fix Source restore state NPE (#2878)

laglangyue committed:
[Improve][Connector-v2-Fake]Supports direct definition of data values(row) (#2839)

* [Improve][Connector-v2]Supports direct definition of data values(row)

ashulin reviewed (commented):
Hey, I have a different idea about Metrics.
For various engines, they already have metrics and metric reporters, and engine users have also managed metrics based on their metrics reporters. 
Seatunnel just submits an engine job, and it is unreasonable for users to do additional metrics management.
My opinion is that seatunnel provides unified metrics for the connector-v2 api, and then adds the metrics of seatunnel to the existing reporter of the engine. 
For older versions of the connector, the engine's own metrics are used directly.
The metrics reporter of seatunnel may only be used for the engine of seatunnel itself.

Hisoka-X commented:
>My opinion is that seatunnel provides unified metrics for the connector-v2 api, and then adds the metrics of seatunnel to the existing reporter of the engine.  

This is a feature will do next.

> For older versions of the connector, the engine's own metrics are used directly.

There alway can use engine own metrics both old version connector and connector-v2

> The metrics reporter of seatunnel may only be used for the engine of seatunnel itself.

At now, our reporter can support flink/spark/seatunnel with a little work. It is not conflict with their own reporter.

lvlv-feifei committed:
seatunnel-prometheus update

lvlv-feifei committed:
Merge remote-tracking branch 'origin/dev' into dev

lvlv-feifei committed:
Merge branch 'apache:dev' into dev

hailin0 commented on seatunnel-metrics/seatunnel-metrics-flink/src/main/java/org/apache/seatunnel/metrics/flink/SeatunnelMetricReporter.java:
```diff
@@ -0,0 +1,111 @@
+            histogramsIndex.put(new SimpleHistogram(key.getCount(), key.getStatistics().getMin(), key.getStatistics().getMax(), key.getStatistics().getStdDev(), key.getStatistics().getMean(), quantile), metric.getValue());
+        }
+        //todo handle user config
+        reporter = new PrometheusPushGatewayReporter(jobName, host, port);
```
Moved outside the `SeatunnelMetricReporter#report()` method?

hailin0 commented on seatunnel-metrics/seatunnel-metrics-prometheus/src/main/java/org/apache/seatunnel/metrics/prometheus/PrometheusPushGatewayReporter.java:
```diff
@@ -0,0 +1,235 @@
+    @Override
+    public PrometheusPushGatewayReporter open() {
+        //todo Handle user config
+        return new PrometheusPushGatewayReporter("flink_prometheus_job", "localhost", DEFAULT_PORT);
```
Why create `PrometheusPushGatewayReporter` object again?

hailin0 commented on seatunnel-metrics/seatunnel-metrics-core/src/main/java/org/apache/seatunnel/metrics/core/reporter/MetricReporter.java:
```diff
@@ -0,0 +1,22 @@
+
+/** Reporters are used to export seatunnel {@link Metric Metrics} to an external backend. */
+public interface MetricReporter {
+    MetricReporter open();
```
Why return `MetricReporter` object? this is object method

hailin0 commented on seatunnel-metrics/seatunnel-metrics-spark/src/main/java/org/apache/seatunnel/metrics/spark/SeatunnelMetricSink.scala:
```diff
@@ -0,0 +1,219 @@
+            }),
+          newMetricInfo(metricName, dimensionKeys, dimensionValues))
+      }
+      val reporter = new PrometheusPushGatewayReporter(pollJobName, pollHost, pollPort)
```
as above

hailin0 commented:
check file license header

lvlv-feifei committed:
seatunnel-prometheus update

lvlv-feifei committed:
seatunnel-prometheus update

lvlv-feifei committed:
Merge remote-tracking branch 'origin/dev' into dev

# Conflicts:
#	seatunnel-e2e/seatunnel-spark-connector-v2-e2e/connector-file-spark-e2e/src/test/resources/file/fakesource_to_local_orc.conf

lvlv-feifei committed:
1. Seatunnel unified configuration naming
2. Use reflection to automate assembly
3. Modify the flink/spark startup function
4. Try packaging configuration (todo)

Hisoka-X reviewed (approved):
LGTM

Hisoka-X commented:
We merge it into `st-metrics` branch, should have long term to develop it before ready.
